src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/EnumerableExtensions.cs

```diff
@@ -59,4 +69,446 @@ public static bool TryGetCount<T>(this IEnumerable sequence, out int count)
         count = 0;
         return false;
     }
+
+    /// <summary>
+    ///  Copies the contents of the sequence to a destination <see cref="Span{T}"/>.
+    /// </summary>
+    /// <typeparam name="T">The type of elements in the sequence.</typeparam>
+    /// <param name="sequence">The sequence to copy items from.</param>
+    /// <param name="destination">The span to copy items into.</param>
+    /// <exception cref="ArgumentException">
+    ///  The destination span is shorter than the source sequence.
+    /// </exception>
+    public static void CopyTo<T>(this IEnumerable<T> sequence, Span<T> destination)
+    {
+        // Check a couple of common cases.
+        switch (sequence)
+        {
+            // We specifically test ImmutableArray<T> to avoid boxing it as an IReadOnlyList.
+            case ImmutableArray<T> array:
+                ArgHelper.ThrowIfDestinationTooShort(destination, array.Length);
+                array.CopyTo(destination);
+                break;
+
+            // HashSet<T> has special enumerator and doesn't implement IReadOnlyList<T>
+            case HashSet<T> set:
+                set.CopyTo(destination);
+                break;
+
+            case IReadOnlyList<T> list:
+                list.CopyTo(destination);
+                break;
+        }
```

Comment:
All of these cases need `return` instead of `break`, right?

Reply:
Yes -- thanks! Can't believe I missed that!